### PR TITLE
Prepare for context argument changes in Prometheus

### PR DESCRIPTION
**Pull request by aknuds1**
#### What this PR does
Prepare for context argument changes (some functions gain a context argument, while others lose it) in Prometheus, by adding a context argument to functions that will need it.

This PR should cause no behavioral changes, it is only supposed to reduce the number of changes required for [updating our Prometheus dependency](https://github.com/grafana/mimir/pull/6085). 

#### Checklist

- [x] Tests updated
- [na] Documentation added
- [na] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storegateway/series_refs.go [hunk 1]
```diff
@@ -955,7 +955,7 @@ func (s *loadingSeriesChunkRefsSetIterator) Next() bool {
 	}
 	defer symbolizedSet.release() // We only retain the slices of chunk ranges from this set. These are still not pooled, and it's ok to retain them.
 
-	nextSet, err := s.stringifiedSet(symbolizedSet)
+	nextSet, err := s.stringifiedSet(context.TODO(), symbolizedSet)
```
Comment (aknuds1) [inline pkg/storegateway/series_refs.go:958]: Adding a context argument to the `Next` method causes a domino effect of other iterator types that'll need the same change, so I opted for `context.TODO()` for now. I'm afraid the change set will be too big otherwise.

File: pkg/querier/blocks_store_queryable.go [hunk 2]
```diff
@@ -342,7 +342,7 @@ type blocksStoreQuerier struct {
 // Select implements storage.Querier interface.
 // The bool passed is ignored because the series is always sorted.
 func (q *blocksStoreQuerier) Select(_ bool, sp *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
-	return q.selectSorted(sp, matchers...)
+	return q.selectSorted(q.ctx, sp, matchers...)
```
Comment (krajorama) [inline pkg/querier/blocks_store_queryable.go:345]: Why is this needed? Doesn't seem to add anything

Comment (krajorama) [reply pkg/querier/blocks_store_queryable.go:345]: discussed IRL

File: pkg/storegateway/series_refs.go (hunk 1 above)
Comment (krajorama) [reply pkg/storegateway/series_refs.go:958]: Could you use `s.ctx` ?

File: pkg/ingester/ingester_test.go [hunk 3]
```diff
@@ -4219,6 +4221,8 @@ func getWALReplayConcurrencyFromTSDBHeadOptions(userTSDB *userTSDB) int {
 }
 
 func TestIngester_shipBlocks(t *testing.T) {
+	ctx := context.Background()
```
Comment (dimitarvdimitrov) [inline pkg/ingester/ingester_test.go:4224]: shipBlocks further down uses context.Backnground(), should we change it to use ctx now?

File: pkg/ingester/ingester_test.go [hunk 4]
```diff
@@ -4460,7 +4464,7 @@ func TestIngester_idleCloseEmptyTSDB(t *testing.T) {
 		return i.lifecycler.HealthyInstancesCount()
 	})
 
-	db, err := i.getOrCreateTSDB(userID, true)
+	db, err := i.getOrCreateTSDB(ctx, userID, true)
```
Comment (dimitarvdimitrov) [inline pkg/ingester/ingester_test.go:4467]: somewhat out of scope, but compactBlocks and shipBlocks further down use context.Backnground(), should we change them to use ctx now?

File: pkg/storage/tsdb/block/index.go [hunk 5]
```diff
@@ -396,14 +396,14 @@ func GatherBlockHealthStats(logger log.Logger, blockDir string, minTime, maxTime
...
-// Repair open the block with given id in dir and creates a new one with fixed data.
+// Repair opens the block with given id in dir and creates a new one with fixed data.
 // It:
 // - removes out of order duplicates
 // - all "complete" outsiders (they will not accessed anyway)
 // - removes all near "complete" outside chunks introduced by https://github.com/prometheus/tsdb/issues/347.
 // Fixable inconsistencies are resolved in the new block.
 // TODO(bplotka): https://github.com/thanos-io/thanos/issues/378.
-func Repair(logger log.Logger, dir string, id ulid.ULID, source SourceType, ignoreChkFns ...ignoreFnType) (resid ulid.ULID, err error) {
+func Repair(ctx context.Context, logger log.Logger, dir string, id ulid.ULID, source SourceType, ignoreChkFns ...ignoreFnType) (resid ulid.ULID, err error) {
```
[the remark this answers is not in the document]
Comment (aknuds1) [reply pkg/storage/tsdb/block/index.go:406]: Fixed, thanks.

File: pkg/querier/tenantfederation/merge_queryable.go [hunk 6]
```diff
@@ -166,7 +166,7 @@ func (m *mergeQuerier) LabelValues(name string, matchers ...*labels.Matcher) ([]
 		name = m.idLabelName
 	}
 
-	return m.mergeDistinctStringSliceWithTenants(func(ctx context.Context, q storage.Querier) ([]string, storage.Warnings, error) {
+	return m.mergeDistinctStringSliceWithTenants(m.ctx, func(ctx context.Context, q storage.Querier) ([]string, storage.Warnings, error) {
```
[the remark this answers is not in the document]
Comment (aknuds1) [reply pkg/querier/tenantfederation/merge_queryable.go]: Thanks, done.

File: pkg/ingester/ingester_test.go (hunk 4 above)
Comment (aknuds1) [reply pkg/ingester/ingester_test.go:4467]: Thanks, done.

File: pkg/ingester/ingester_test.go (hunk 3 above)
Comment (aknuds1) [reply pkg/ingester/ingester_test.go:4224]: Thanks, done.

File: pkg/storegateway/series_refs.go (hunk 1 above)
Comment (aknuds1) [reply pkg/storegateway/series_refs.go]: Switched to `s.ctx`.

**Review by krajorama — APPROVED** [review]
LGTM

**Comment by aknuds1** [comment]
Thanks for the careful reviews @krajorama and @dimitarvdimitrov!

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM, thanks for addressing my comments

File: pkg/ingester/ingester.go [hunk 7]
```diff
@@ -830,7 +830,7 @@ func (i *Ingester) PushWithCleanup(ctx context.Context, pushReq *push.Request) (
 		return &mimirpb.WriteResponse{}, nil
 	}
 
-	db, err := i.getOrCreateTSDB(userID, false)
+	db, err := i.getOrCreateTSDB(ctx, userID, false)
```
Comment (pracucci) [inline pkg/ingester/ingester.go:833]: We're passing the request context to `getOrCreateTSDB()`. I don't know what it will be used for, but if the context will be retained by TSDB then I think this is a mistake. We shouldn't pass any request-specific context to any place where the context will be retained.

Comment (aknuds1) [reply pkg/ingester/ingester.go:833]: Thanks for the feedback @pracucci. The context shouldn't be retained, but I'm looking into whether it's a bad idea to pass it down into TSDB.
